feat(resource-types): add resource-types (#76)
aep-dev/aep.dev: aep/general/0004/aep.md

# Resource types

Most APIs expose _resources_ (their primary nouns) which users are able to
create, retrieve, and manipulate. APIs are allowed to name their resource types
as they see fit, and are only required to ensure uniqueness within that API.
This means that it is possible (and often desirable) for different APIs to use
the same type name. For example, a Memcache and Redis API would both want to
use `Instance` as a type name.

When mapping the relationships between APIs and their resources, however, it
becomes important to have a single, globally-unique type name. Additionally,
tools such as Kubernetes or GraphQL interact with APIs from multiple providers.

## Guidance

APIs **must** define a resource type for each resource in the API, according to

> the following pattern: `{API Name}/{Type}`. The type name:

Author: Do you have a suggestion? I at least called it "Type name" to not feel as recursive.

> the following pattern: `{API Name}/{Type}`. The type name:

Author: added type name, PTAL.

- **must** Start with an uppercase letter.
- **must** Only contain alphanumeric characters.
- **must** Be of the singular form of the noun.
- **must** Use PascalCase (UpperCamelCase).
- For Kubernetes, the type name **must** match the [object][] name.

> - For Protobuf, the type name **must** match the name of the protobuf message.

Author: The guidance below talks about the usage of a proto annotation, which doesn't necessarily have to match. So I think it's good to be explicit here.

Added OAS guidance.

> - For Protobuf, the type name **must** match the name of the protobuf message.

Reviewer: Gotcha.  Not marking this resolved because there's still a casing suggestion that isn't committed.

### Examples

Examples of resource types include:

- `pubsub.googleapis.com/Topic`
- `pubsub.googleapis.com/Subscription`
- `spanner.googleapis.com/Database`
- `spanner.googleapis.com/Instance`

> - `networking.istio.io/Instance`

Author: done.

### Annotating resource types

APIs **must** annotate the resource types for each resource in the API using

> the [`google.api.resource`][resource] annotation:

Author: I'd rather focus on proto for now and see how things like in OAS later at the end - primarily because I have the aepc repository which just uses what grpc-proxy translates things as.

How about an issue on it for later?

> the [`google.api.resource`][resource] annotation:

Reviewer: Hmm...we use proto at Roblox so it doesn't affect me in that sense, but I wonder how our OAS-aligned members will feel about OAS being effectively a secondary consideration. :/

> the [`google.api.resource`][resource] annotation:

Author: I'll leave it open, we can raise concerns there. Personally I'd like to see if we can find a better way to have dual support for OAS / proto than spelling out every annotation on each (although maybe that's the way it has to be).

> the [`google.api.resource`][resource] annotation:

Author: added OAS annotations.

```proto
// A representation of a Pub/Sub topic.
message Topic {
  option (google.api.resource) = {
    type: "pubsub.googleapis.com/Topic"
    pattern: "projects/{project}/topics/{topic}"
    singular: "topic"
    plural: "topics"
  };

  // Name and other fields...
}
```

> - Patterns **must** correspond to the [resource path][resource-paths].

Author: resource paths are the spec, correct? If we define a spec, I'd suggest it is defined in resource path and we link to there.

> - Patterns **must** correspond to the [resource path][resource-paths].

Reviewer: Well...what I meant is that the pattern uses curly-brace variables that aren't defined by the resource-paths AEP.

> - Patterns **must** correspond to the [resource path][resource-paths].

Author: ack, yeah I can define it. Let's resolve https://github.com/aep-dev/aep.dev/pull/76/files#r1398519206 then I'll add it.

> - Patterns **must** correspond to the [resource path][resource-paths].

Author: okay, I rewrote this section, PTAL. added a grammar and more specific guidance, let's talk there.

- Pattern variables (the segments within braces) **must** use `snake_case`, and
  **must not** use an `_id` suffix.

> - Pattern variables **must** conform to the format `[a-z][_a-z0-9]*[a-z0-9]`.

Author: That's not invalid, I believe - In practice it maps to a protobuf field name, and I think a_____9 is valid?

> - Pattern variables **must** conform to the format `[a-z][_a-z0-9]*[a-z0-9]`.

Reviewer: I agree that's a valid field name, but pattern variables don't correspond to a field name, they correspond to a the name of a resource type (which should be in PascalCase, and converting PascalCase to snake_case never results in adjacent underscores).

> - Pattern variables **must** conform to the format `[a-z][_a-z0-9]*[a-z0-9]`.

Author: corresponding to a resource type seems reasonable - it may also avoid the issue where it's hard to figure our the parent resource types in the pattern, since the collection names can be short-hand for the resource (that clause in the resource path about https://aep.dev/122#nested-collections)

In that case we would have stuff like:

```/users/{user}/events/{userEvent}```

is that ok with you?

> - Pattern variables **must** conform to the format `[a-z][_a-z0-9]*[a-z0-9]`.

Author: Actually I was thinking through this a little more - and I think that in general, patterns a somewhat flawed - they force you to re-enumerate the resource paths of parents, or worse have a parent resource path that doesn't match.

Really the only thing a sub-resource can do is add a suffix to a specific parent. So really, I think we should do something like:

```
google.api.resource = {
    type: "user-event"
    singular: "user-event"
    plural: "user-events"
    parents = [ 
        type: "user",
        path_suffix = "/events/{user-event}"
    ]
}
```

does that make sense? I think we'd have to come up with our own resource annotation to support it, but it allows for generation of correct, fully qualified resource paths.

> - Pattern variables **must** conform to the format `[a-z][_a-z0-9]*[a-z0-9]`.

Author: the pattern annotation is used by the api-linter, but it's not used for gRPC-HTTP bindings at all.

My opinion is that we should fork the annotation here so it's a bit more semantically clear, and work from parent references. But @rofrankel noted in our last meeting that having all patterns enumerated is nice from a documentation perspective.

We've agreed upon keeping the existing annotation as-is, but introducing a more standard resource definition syntax into the AEPs.

I might try to just fix this AEP to work without it though (e.g. documenting the pattern syntax as suggested) so that we can get this pretty critical AEP in while we debate resource syntax.

> - Pattern variables **must** conform to the format `[a-z][_a-z0-9]*[a-z0-9]`.

Author: fixed the aep as-is for now, I'll work on adding the resource definition at a future time.

- Pattern variables **must** be unique within any given pattern. (e.g.
  `projects/{abc}/topics/{abc}` is invalid; this is usually a natural corollary
  of collection identifiers being unique within a pattern.)

> - Singular **must** be the lower camel case of the type.

Author: Singular variable - I'll update to clarify. 

TLDR on the snake / camel discrepancy - yeah they should match, probably snake, but:

There is a distinction between pattern variables and singular types - they don't necessarily have to match to work: pattern variables match to proto fields in Google's API frameworks, while the singular / plural is used more to match with collection names. Of course, with the advent of shortening the collection name, we can't even rely on singular / plural being used in the path at all anymore.

We should normalize it, but I think it'd be good to clarify what sort of consistency users should expect based on this - with resource path singletons, API paths, and no guarantee that path segments match collection names, there's basically no technical reason for any of this to match and clients have to assume none of them do.

> - Singular **must** be the lower camel case of the type.

Reviewer: Are you thinking of google.api.http routes (e.g. `{parent=...}` rather than google.api.resource patterns?

> - Singular **must** be the lower camel case of the type.

Author: I'm referring to the section above.

> - Singular **must** be the lower camel case of the type.

Author: updated the AEP quite a bit to be clearer here, including an grammar. Please review and leave any comments on the new text.

  - Pattern variables **must** be the singular form of the resource type e.g. a
    pattern variable representing a `Topic` resource ID is named `{topic}`.

> - Plural **must** be the lower camel case plural of the singular.

Author: value in the annotation - I'll update.

> - Plural **must** be the lower camel case plural of the singular.

Reviewer: I don't understand the current phrasing (and also, should it be snake_case rather than lowerCamelCase)?

Separately, elsewhere in this file it's spelled "lowerCamelCase" rather than "lower camel case".

> - Plural **must** be the lower camel case plural of the singular.

Author: so after some thinking, I've updated this section to use kebab-case to be consistent with things like resource path. 

Why do you think it should be snake_case? I was thinking kebab-case to remove yet another place where there's a mostly similar string that has to be transformed somehow.

  - Pattern collection identifier segments **must** match the plural of the

>     resources, except in the case of [nested collections][].

Author: ack, you're right.

#### Pattern uniqueness

> When multiple patterns are defined within a resource, these patterns **must**

Author: I think it's a good suggestion! added.

> When multiple patterns are defined within a resource, these patterns **must**

Reviewer: Did this ever get updated?

> When multiple patterns are defined within a resource, these patterns **must**

Author: yes, I think you may be looking at an outdated diff. Although I rewrote this sentence anyway.

> be mutually unique, where uniqueness is defined as being by-character identical

Author: nope, removed.

once all resource ID path segments have been removed, leaving all `/`
separators.

Therefore the following two patterns **must not** be defined within the same
resource:

- `user/{user}`

> - `user/{user_part_1}~{user_part_2}`

Author: I think I get your point - the paragraph is kind of roundabout?

I tried to focus specifically on the fact that it has to ensure unique matches - PTAL.

> - `user/{user_part_1}~{user_part_2}`

Reviewer: Seems better, thanks.  I added a couple superficial suggestions on that text, but resolving this comment.

## Rationale

### Singular and Plural

Well-defined singular and plurals of a resource enable clients to determine the
proper name to use in code and documentation.

lowerCamelCase can be translated into other common forms of a resource path

> such as UpperCamelCase and snake_case.

Reviewer: I don't know exactly what this is the rationale for...and all the string cases can be converted to the others, right?

Author: I think the main issue was acronyms - e.g. if you have a resource name APIMethod for example - how would one know that should be translated to API_method?

Although I guess should we just have the type be in kebab-case, to match the resource path?

Author: updated to kebab-case.

# Resource types

Most APIs expose _resources_ (their primary nouns) which users are able to
create, retrieve, and manipulate. APIs are allowed to name their resource types
as they see fit, and are only required to ensure uniqueness within that API.
This means that it is possible (and often desirable) for different APIs to use
the same type name. For example, a Memcache and Redis API would both want to
use `Instance` as a type name.

When mapping the relationships between APIs and their resources, however, it
becomes important to have a single, globally-unique type name. Additionally,
tools such as Kubernetes or GraphQL interact with APIs from multiple providers.

## Guidance

APIs **must** define a resource type for each resource in the API, according to
the following pattern: `{API Name}/{Type Name}`. The type name:

- **must** Start with an uppercase letter.

> - **must** Only contain alphanumeric characters.

Reviewer 2: ```suggestion
- **must** Only contain ASCII alphanumeric characters.
```

```proto
// A representation of a Pub/Sub topic.
message Topic {
  option (google.api.resource) = {
    type: "pubsub.googleapis.com/Topic"
    pattern: "projects/{project}/topics/{topic}"
    singular: "topic"
    plural: "topics"
  };

  // Name and other fields...
}
```

- Patterns **must** correspond to the [resource path][resource-paths].
- Pattern variables (the segments within braces) **must** use `snake_case`, and
  **must not** use an `_id` suffix.
- Pattern variables **must** conform to the format `[a-z][_a-z0-9]*[a-z0-9]`.
- Pattern variables **must** be unique within any given pattern. (e.g.
  `projects/{abc}/topics/{abc}` is invalid; this is usually a natural corollary
  of collection identifiers being unique within a pattern.)
- Singular **must** be the lower camel case of the type.
  - Pattern variables **must** be the singular form of the resource type e.g. a
    pattern variable representing a `Topic` resource ID is named `{topic}`.
- Plural **must** be the lower camel case plural of the singular.
  - Pattern collection identifier segments **must** match the plural of the
    resource, except in the case of [nested collections][].

#### Pattern uniqueness

> If multiple patterns are defined within a resource, the patterns defined must
> not overlap in the set of resource paths that they can match. In other words, a

Reviewer: ```suggestion
If multiple patterns are defined within a resource, the patterns defined **must
not** overlap in the set of resource paths that they can match. In other words, a
```

resource path may match at most one of the patterns.

> Therefore the following two patterns **must not** be defined within the same

Reviewer: ```suggestion
For example, the following two patterns would not be valid for the same
```

`**must not**` in an example feels a bit off to me, so suggesting slightly different phrasing.

APIs **must** define a resource type for each resource in the API, according to
the following pattern: `{API Name}/{Type Name}`. The type name:

- **must** Only contain ASCII alphanumeric characters.
- **must** Start with a lowercase letter.
- **must** Be of the singular form of the noun.
- **must** Use kebab case.
- For Kubernetes, the type name when converted to upper camel case **must**
  match the [object][] name.
- For OpenAPI, the type name when converted to upper camel case **must** match
  the name of the schema representing the object.
- For protobuf, the type name when converted to upper camel case **must** match
  the name of the protobuf message.

### Examples

Examples of resource types include:

- `networking.istio.io/instance`
- `pubsub.example.com/topic`
- `pubsub.example.com/subscription`
- `spanner.example.com/database`
- `spanner.example.com/instance`

> - `user.example.com/user-event`

Reviewer: What about the `apis.example.com/users/user-event` form?  I don't want to argue for that to be _the_ canonical form, but I do think it should be supported.  Otherwise we are prescribing that everyone has to use a subdomain for each API, which raises the barrier to entry (and would have been a dealbreaker for Roblox, for example).

Author: We do have an outstanding issue to pick up this discussion at some point: https://github.com/aep-dev/aep.dev/issues/74#issue-2000272320.

I'll update the example as suggested, but my main concern is we have to go put these types of examples everywhere - ideally we'd consolidate discussion of a single concept to a single AEP so we don't have to go rewrite a ton of examples as things change.

### Annotating resource types

APIs **must** annotate the resource types for each resource in the API

{% tab proto %}

For protobuf, use the [`google.api.resource`][resource] annotation:

```proto
// A representation of a user event.
message Topic {
  option (google.api.resource) = {
    type: "user.example.com/user-event"
    singular: "user-event"
    plural: "user-events"
    // define one or more patterns, e.g. if a resource has more than one parent.
    pattern: "projects/{project}/user-events/{user-event}"
    pattern: "folder/{folder}/user-events/{user-event}"
    pattern: "users/{user}/events/{user-event}"
  };

  // Name and other fields...
```

{% tab oas %}

For OpenAPI 3.0, use the `x-aep-resource` extension:

```json
{
  "type": "object",

>   "x-aep-resource": {

Reviewer: Is this extension defined anywhere e.g. as a JSON Schema?  (Does it need to be?)

Author: it's a good question! It definitely should be documented somewhere.

The question is where. I'm thinking maybe one of:

1. an AEP
2. in something like your aep repository

Personally I err toward something a little more fully encapsulated, and think we should document it in this repo. But I'm flexible. Definitely a topic we can discuss as a group?

Author: schema added, ptal.

Reviewer 3: I'm not sure I understand the goal here from the snippets of conversation. It might help me to have an explanation of what `x-aep-resource` is meant to capture? I suspect there's a limitation in the OpenAPI that you're trying to compensate for?

    "singular": "user-event",
    "plural": "user-events",
    "patterns": [
      "projects/{project}/user-events/{user-event}",
      "folder/{folder}/user-events/{user-event}",
      "users/{user}/events/{user-event}"
    ]
  }
}
```

{% endtabs %}

> - The `singular` field **must** match the type name.

Reviewer: ```suggestion
- The `singular` field **must** be the kebab-case singular type name.
```

Well-defined singular and plurals of a resource enable clients to determine the
proper name to use in code and documentation.

> google.aip.dev defines the resource type to use upper camel case, while aeps

Reviewer: ```suggestion
google.aip.dev defines the resource type to use UpperCamelCase, while aeps
```

> google.aip.dev defines the resource type to use upper camel case, while aeps
> define the kebab-case. This is to enforce better consistency in the

Reviewer: ```suggestion
google.aip.dev uses upper camel case for resource types, while aep.dev
uses kebab-case. This is to enforce better consistency in the
```
